### PR TITLE
eye speye deprecated methods

### PR DESCRIPTION
**Pull request by Sacha0**
Towards https://github.com/JuliaLang/LinearAlgebra.jl/issues/454, this pull request explores deprecating `speye`. Please see observations from sketching this pull request in https://github.com/JuliaLang/LinearAlgebra.jl/issues/454. Best!

### CODE REVIEW

File: base/deprecated.jl [hunk 1]
```diff
@@ -2056,6 +2056,13 @@ end
 # deprecate bits to bitstring (#24263, #24281)
 @deprecate bits bitstring
 
+# deprecate speye
+@deprecate speye(n::Integer) sparse(1.0I, n)
+@deprecate speye(m::Integer, n::Integer) sparse(1.0I, m, n)
+@deprecate speye(::Type{T}, n::Integer) where {T} = sparse(UniformScaling{T}(one(T)), n)
```
Comment (ararslan) [inline base/deprecated.jl]: Misplaced `=`?

Comment (Sacha0) [reply base/deprecated.jl]: Good catch! Fixed. Thanks Alex! :)

**Comment by Sacha0** [comment]
As @fredrikekre prompted in https://github.com/JuliaLang/julia/issues/23156#issuecomment-339877052 / https://github.com/JuliaLang/julia/issues/23156#issuecomment-340028507 , we might consider simultaneously deprecating `sparse(I, m, n)` (as opposed to `sparse(I, n)`). Best!

**Comment by Sacha0** [comment]
Fixed a few mistakes, added news, and removed the WIP label. Downstream actions include introducing a `SparseMatrixCSC(I, m, n)` constructor and potentially deprecating `sparse(I, m, n)`, though perhaps there's no need. Best!

**Comment by ViralBShah** [comment]
I think `sparse(I, m, n)` is useful. I can't recollect my use cases though.

**Comment by Sacha0** [comment]
Rebased and revised with #24372 and #24396 in. Pending CI, this pull request should be in shape. Best!

File: base/deprecated.jl [hunk 2]
```diff
@@ -2062,6 +2062,57 @@ end
 # deprecate bits to bitstring (#24263, #24281)
 @deprecate bits bitstring
 
+# deprecate speye
```
Comment (fredrikekre) [inline base/deprecated.jl]: Should `export speye` to avoid #24325 again :)

Comment (Sacha0) [reply base/deprecated.jl]: Good catch! One day I will remember :). Also fixed. Thanks! :)

File: base/deprecated.jl [hunk 3]
```diff
@@ -2062,6 +2062,58 @@ end
...
+export speye
+function speye(n::Integer)
+    depwarn(string("`speye(n::Integer)` has been deprecated in favor of `I`, `sparse`, and ",
+                    "`SparseMatrixCSC` constructor methods. For a direct replacement, consider ",
+                    "`sparse(1.0I, n)`, `SparseMatrixCSC(1.0I, n)`, or `SparseMatrixCSC{Float64}(I, n)`. ",
+                    "If `Float64` element type is not necessary, consider the shorter `sparse(I, n)` ",
+                    "or `SparseMatrixCSC(I, n)` (with default `eltype(I)` of `Bool`)."), :speye)
+    return speye(1.0I, n)
```
Comment (JeffBezanson) [inline base/deprecated.jl]: `sparse(1.0I, n))` ?

**Comment by Sacha0** [comment]
Plan per triage: Remove the single-integer versions of `sparse(I, ...)` and `SparseMatrixCSC(I, ...)` and merge. Best!

File: base/deprecated.jl (hunk 3 above)
Comment (Sacha0) [reply base/deprecated.jl]: Good catch! Fixed on push. Thanks!

**Comment by Sacha0** [comment]
Revised in accord with triage, removing the single-integer versions of `sparse(I, ...)` and `SparseMatrixCSC(I, ...)`. Best!

**Comment by Sacha0** [comment]
Ref. https://github.com/JuliaLang/julia/pull/24438#issuecomment-341999990. Best!

**Comment by Sacha0** [comment]
Rebased. Absent objections or requests for time, I plan to merge this pull request tomorrow. Best!

**Comment by Sacha0** [comment]
Thanks all! :)

File: base/deprecated.jl [hunk 4]
```diff
@@ -2076,6 +2076,55 @@ end
...
+    return sparse(1.0I, m, n)
+end
+function speye(::Type{T}, n::Integer) where T
+    depwarn(string("`speye(T, n::Integer)` has been deprecated in favor of `I`, `sparse`, and ",
+                    "`SparseMatrixCSC` constructor methods. For a direct replacement, consider ",
+                    "`sparse(T(1)I, n, n)` if `T` is concrete or `SparseMatrixCSC{T}(I, n, n)` ",
+                    "if `T` is either concrete or abstract. If element type `T` is not necessary, ",
+                    "consider the shorter `sparse(I, n, n)` or `SparseMatrixCSC(I, n, n)` ",
+                    "(with default `eltype(I)` of `Bool`)."), :speye)
+    return SparseMatrixCSC{T}(I, n)
```
Comment (martinholters) [inline base/deprecated.jl]: `SparseMatrixCSC{T}(I, n, n)`

Currently, 
```julia
julia> speye(Float64, 4)
WARNING: `speye(T, n::Integer)` has been deprecated in favor of `I`, `sparse`, and `SparseMatrixCSC` constructor methods. For a direct replacement, consider `sparse(T(1)I, n, n)` if `T` is concrete or `SparseMatrixCSC{T}(I, n, n)` if `T` is either concrete or abstract. If element type `T` is not necessary, consider the shorter `sparse(I, n, n)` or `SparseMatrixCSC(I, n, n)` (with default `eltype(I)` of `Bool`).
Stacktrace:
 [1] depwarn(::String, ::Symbol) at ./deprecated.jl:68
 [2] speye(::Type{Float64}, ::Int64) at ./deprecated.jl:2145
 [3] top-level scope
 [4] eval(::Module, ::Expr) at ./repl/REPL.jl:3
 [5] eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./repl/REPL.jl:69
 [6] macro expansion at ./repl/REPL.jl:100 [inlined]
 [7] (::getfield(Base.REPL, Symbol("##1#2")){Base.REPL.REPLBackend})() at ./event.jl:95
in expression starting at no file:0
ERROR: MethodError: no method matching SparseMatrixCSC{Float64,Ti} where Ti<:Integer(::UniformScaling{Bool}, ::Int64)
Stacktrace:
 [1] speye(::Type{Float64}, ::Int64) at ./deprecated.jl:2151
 [2] top-level scope
```

Comment (Sacha0) [reply base/deprecated.jl]: Thanks `@martinholters`! :)

**Pull request merged**